libp2p/autotls/acme.nim

```diff
@@ -0,0 +1,316 @@
+import options, json, base64
+from times import DateTime, parse
+import chronos/apps/http/httpclient, jwt, results, bearssl/pem
+
+import ./utils
+import ../crypto/crypto
+import ../crypto/rsa
+import ../transports/tls/certificate_ffi
+import ../transports/tls/certificate
+
+const
+  LetsEncryptURL* = "https://acme-v02.api.letsencrypt.org"
+  LetsEncryptURLStaging* = "https://acme-staging-v02.api.letsencrypt.org"
+  Alg = "RS256"
+  DefaultChalCompletedRetries = 10
+  DefaultChalCompletedRetryTime = 1.seconds
+  DefaultFinalizeRetries = 10
+  DefaultFinalizeRetryTime = 1.seconds
+  DefaultRandStringSize = 256
+
+type ACMEAccount* = object
+  status*: Opt[string]
+  contact*: Opt[seq[string]]
+  key*: KeyPair
+  session*: HttpSessionRef
+  kid*: Opt[string]
+  directory: JsonNode
+  acmeServerURL: string
+
+proc new*(
+    T: typedesc[ACMEAccount],
+    key: KeyPair,
+    status: Opt[string] = Opt.none(string),
+    contact: Opt[seq[string]] = Opt.none(seq[string]),
+    kid: Opt[string] = Opt.none(string),
+    acmeServerURL: string = LetsEncryptURL,
+): Future[ref ACMEAccount] {.async: (raises: [ACMEError, CancelledError]).} =
+  let session = HttpSessionRef.new()
+  var directory: JsonNode
```

Comment:
why `JsonNode` are used? why not parse it to appropriate model right away and use model. it gives better type safety then accessing individual properties using `getJSONField`, etc..